core/src/test/java/org/infinispan/xsite/irac/IracTombstoneUnitTest.java

```diff
@@ -0,0 +1,265 @@
+      return im;
+   }
+
+   private static ScheduledExecutorService createScheduledExecutorService(Queue<? super RunnableData> queue) {
+      ScheduledExecutorService executorService = Mockito.mock(ScheduledExecutorService.class);
+      Mockito.when(executorService.schedule(ArgumentMatchers.any(Runnable.class), ArgumentMatchers.anyLong(), ArgumentMatchers.any())).thenAnswer(invocationOnMock -> {
+         queue.add(new RunnableData(invocationOnMock.getArgument(0), invocationOnMock.getArgument(1)));
+         return null;
+      });
+      return executorService;
+   }
+
+   private static IracMetadata createIracMetadata() {
+      return Mockito.mock(IracMetadata.class);
+   }
+
+   private static DefaultIracTombstoneManager createIracTombstoneManager(Queue<? super RunnableData> queue, int targetSize, long maxDelay, AtomicBoolean keep) {
+      DefaultIracTombstoneManager manager = new DefaultIracTombstoneManager(createConfiguration(targetSize, maxDelay));
+      TestingUtil.inject(manager,
+            createDistributionManager(),
+            createTakeOfflineManager(),
+            createIracManager(keep),
+            new WithinThreadExecutor(),
+            createScheduledExecutorService(queue),
+            createCommandFactory(),
+            createRpcManager());
+      return manager;
+   }
+
+   public void testDelayIncreaseWithNoTombstones() throws InterruptedException {
+      BlockingDeque<RunnableData> queue = new LinkedBlockingDeque<>();
+      DefaultIracTombstoneManager manager = createIracTombstoneManager(queue, 1, 1000, new AtomicBoolean(false));
+      manager.start();
+
+      RunnableData data = queue.poll(10, TimeUnit.SECONDS);
+      assertNotNull(data);
+      assertEquals(500, data.delay);
+
+      // check the max limit
+      for (long expectedDelay : Arrays.asList(707, 841, 917, 958, 979, 989, 994, 997, 998, 999, 999, 999)) {
+         data.runnable.run();
+         data = queue.poll(10, TimeUnit.SECONDS);
+         assertNotNull(data);
+         assertEquals(expectedDelay, data.delay);
+      }
+      manager.stop();
+   }
+
+   public void testDelayConstantAtSameRate() throws InterruptedException {
+      int targetSize = 20;
+      BlockingDeque<RunnableData> queue = new LinkedBlockingDeque<>();
+      DefaultIracTombstoneManager manager = createIracTombstoneManager(queue, targetSize, 2000, new AtomicBoolean(false));
+      manager.start();
+
+      RunnableData data = queue.poll(10, TimeUnit.SECONDS);
+      assertNotNull(data);
+      assertEquals(1000, data.delay);
+
+      IracMetadata metadata = createIracMetadata();
+      insertTombstones(targetSize, manager, metadata);
+
+      data.runnable.run();
+      data = queue.poll(10, TimeUnit.SECONDS);
+      assertNotNull(data);
+      assertEquals(1000, data.delay);
+      manager.stop();
+   }
+
+   public void testDelayConstantAtHigherRate() throws InterruptedException {
```

Comment:
@pruivo one more nit to pick: are `testDelayConstantAtHigherRate` and `testDelayConstantAtLowerRate` named correctly? Because the delay doesn't seem to stay constant to me :)

Reply:
updated